diff --git a/modules/juce_audio_devices/native/juce_Audio_ios.cpp b/modules/juce_audio_devices/native/juce_Audio_ios.cpp
index 709777cec0..14fb21db13 100644
--- a/modules/juce_audio_devices/native/juce_Audio_ios.cpp
+++ b/modules/juce_audio_devices/native/juce_Audio_ios.cpp
@@ -372,9 +372,15 @@ struct iOSAudioIODevice::Pimpl final : public AsyncUpdater
         if (category == AVAudioSessionCategoryPlayAndRecord)
         {
             options |= AVAudioSessionCategoryOptionDefaultToSpeaker
-                     | AVAudioSessionCategoryOptionAllowBluetooth
-                     | AVAudioSessionCategoryOptionAllowAirPlay
-                     | AVAudioSessionCategoryOptionAllowBluetoothA2DP;
+                        | AVAudioSessionCategoryOptionAllowAirPlay
+                        | AVAudioSessionCategoryOptionAllowBluetoothA2DP;
+                      /*| AVAudioSessionCategoryOptionAllowBluetooth */;
+
+            if (@available (iOS 14.5, *))
+                options |= AVAudioSessionCategoryOptionOverrideMutedMicrophoneInterruption;
+
+            if (@available (iOS 10.0, *))
+                options |= AVAudioSessionCategoryOptionAllowBluetoothA2DP;
         }
 
         JUCE_NSERROR_CHECK ([[AVAudioSession sharedInstance] setCategory: category
